auth_saml/models/auth_saml.py

```diff
@@ -1,78 +1,331 @@
+            self.env["ir.config_parameter"].sudo().get_param("web.base.url", "")
         )
-        server.addProviderFromBuffer(
-            lasso.PROVIDER_ROLE_IDP,
-            self.idp_metadata
+
+        for record in self:
+            if isinstance(record.id, models.NewId):
+                record.sp_metadata_url = None
+                continue
+
+            if record.sp_baseurl:
+                base_url = self.sp_baseurl
+            else:
+                base_url = icp_base_url
+
+            qs = urllib.parse.urlencode({"p": record.id, "d": self.env.cr.dbname})
+
+            record.sp_metadata_url = urllib.parse.urljoin(
+                base_url, ("/auth_saml/metadata?%s" % (qs))
+            )
+
+    def _get_config_for_provider(self, base_url=None):
+        """
+        Internal helper to get a configured Saml2Client
+        """
+        self.ensure_one()
+
+        if self.sp_baseurl:
+            base_url = self.sp_baseurl
+
+        if not base_url:
+            base_url = (
+                self.env["ir.config_parameter"].sudo().get_param("web.base.url", "")
+            )
+
+        acs_url = urllib.parse.urljoin(base_url, "/auth_saml/signin")
+
+        keys = self.env["ir.attachment"].search(
+            [
+                ("res_model", "=", self._name),
+                ("res_field", "=", "sp_pem"),
+                ("res_id", "=", self.id),
+            ],
+            limit=1,
         )
-        return lasso.Login(server)
 
-    @api.multi
-    def _get_auth_request(self, state):
-        """build an authentication request and give it back to our client
+        attachment_location = self.env["ir.config_parameter"].sudo().get_param(
+            "ir_attachment.location", "file")
+        if attachment_location != 'file':
+            keys._save_to_file_system()
+        keys_path = self.env["ir.attachment"]._full_path(keys.store_fname)
+
+        settings = {
+            "metadata": {"inline": [self.idp_metadata]},
+            "entityid": self.entity_id,
+            "service": {
+                "sp": {
+                    "endpoints": {
+                        "assertion_consumer_service": [
+                            (acs_url, saml2.BINDING_HTTP_REDIRECT),
+                            (acs_url, saml2.BINDING_HTTP_POST),
+                            (acs_url, saml2.BINDING_HTTP_REDIRECT),
+                            (acs_url, saml2.BINDING_HTTP_POST),
+                        ],
+                    },
+                    "allow_unsolicited": self.idp_allow_unsolicited,
+                    "authn_requests_signed": True,
+                    "logout_requests_signed": True,
+                    "want_assertions_signed": True,
+                    # TODO: as we are not correctly generating metadata, don't require
+                    #  signed responses, since default metadata configurations
+                    #  do not require them
+                    "want_response_signed": False,
```

Comment:
Hi @eLBati can you give a bit more detail on this change? It's something I have working in live, if there's a bug with a particular IDP I can see if I can help out? (for ref I've got a copy working in 12.0 against Duo)